The Changeling armblade now has a spin attack

boy2mantwicethefam opened the pull request:
![514-2891147996](https://github.com/vgstation-coders/vgstation13/assets/41342767/34348588-88eb-483e-99da-a47489eaf382)

- Active for 1.5 seconds
- Makes the changeling move in their current direction at a slower speed, because they are spinning forward
- Automatically hits everyone near the changeling for a total of 3 times
- Cooldown of 30 seconds
- Can't hit people that are laying down
- Biased towards torso, arm and head hits
- Plays a cool sound (thanks @strawberrytan!)
- Cool spinning!
- Anti-cheese features (can't do it while driving, starts off with the 30 second cooldown when freshly formed)

:cl:
 * rscadd: The Changeling's armblade now has a powerful spin attack! Keep your distance!

DrSnips commented:
Oh joy, I can't wait to take 30 x 3 = 90 damage from an item with a 1.5 second cooldown

boy2mantwicethefam commented:
Come on, it's a cool ability...

DrSnips commented:
I think it's cool too, but this basically insta-kills you with these numbers. Why is is 3 times?

MercuriusGH commented:
![firefox_Gb0GLI7AZN](https://github.com/vgstation-coders/vgstation13/assets/141453111/423a8996-d9de-4ca9-a744-d4a34b8eae72)

ARMOR FORM WHEN????????????????

boy2mantwicethefam commented:
It's how it is in Prototype from what I remember, though in Prototype 2 it had a variable number of spins during a charged attack depending on upgrades. it beats the 6 times over 3 seconds that I planned for it before common sense prevailed. The attacks are conventional and can miss, and if there is movement involved not all attacks might land.
I would definitely like to tweak some stuff about it more.
@MercuriusGH Believe it or not TG beat us to it.

MercuriusGH commented:
@boy2mantwicethefam Fuck
Prototype is gigakino and we need more Prototype-based changeling powers

ErrorsWindows commented:
it's about time changeling embraces its true inspiration

jwhitak commented:
I can't wait to kill people in an aoe around me every 30 seconds!

boy2mantwicethefam commented:
The spin attack will now force the changeling to move in their current direction at a slower speed than regular movement.
It will no longer attack people lying down (armblade goes over them) and that includes if they were knocked down by an armblade attack.

boy2mantwicethefam commented:
Some fixes to how rapid the attacks actually were.
Now plays a cool blade wish wosh sound, thanks to @strawberrytan for providing the audio file!
Now targets the torso area with a chance to hit the arms or head instead, same as regular aiming towards the torso.

boy2mantwicethefam commented:
Made the spin more consistent, and the position of the armblade on your arm influences in which direction you'll spin (left or right).
In case the changeling is knocked out of the spin early they will immediately recover from the spin-induced delays instead of being stuck for the rest of what would have been the spin's duration.
And now, demonstration video (with sound)!

https://github.com/vgstation-coders/vgstation13/assets/41342767/5fbb1bbb-3cb7-47d7-a923-d594d29ba7df

boy2mantwicethefam commented:
Fixed the spin direction being incorrect.
The targeting will now properly roll the hit limb for every target instead of always hitting the same limb on all targets.
Converted the variables related to the spin's effects into variables that can be edited by an admin.

Servalfang commented:
JUGGERNAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUUUUUUUUUUUUUUT

ShiftyRail commented on code/game/objects/items/weapons/armblade.dm:
```diff
@@ -15,9 +15,104 @@
+	var/spin_facing = initial_direction //For the purpose of where to spin next
+	var/delay_track = spin_duration //If the action ends prematurely for some reason it will free the changeling of the remaining duration
+
+	visible_message("<span class='sinister'>[user] starts wildly spinning their armblade around!</span>")
```
```suggestion
	visible_message("<span class='sinister'>\The [user] starts wildly spinning their armblade around!</span>")
```

ShiftyRail commented on code/game/objects/items/weapons/armblade.dm:
```diff
@@ -15,9 +15,104 @@
+	var/delay_track = spin_duration //If the action ends prematurely for some reason it will free the changeling of the remaining duration
+
+	visible_message("<span class='sinister'>[user] starts wildly spinning their armblade around!</span>")
+	user.delayNextMove(spin_duration, 1) //Can't move during the spin
```
Rather than this, I would just make it register an event that cancels the spin if you move, so that you're not locking yourself out of control for 1.5 seconds (+/- lag)

ShiftyRail commented on code/game/objects/items/weapons/armblade.dm:
```diff
@@ -15,9 +15,104 @@
+	visible_message("<span class='sinister'>[user] starts wildly spinning their armblade around!</span>")
+	user.delayNextMove(spin_duration, 1) //Can't move during the spin
+	user.delayNextAttack(spin_duration, 1) //Can't attack extra times during the spin
+	for(var/i = 0, i < spin_duration, i++)
```
```suggestion
	for(var/i = 0 to spin_duration)
```

ShiftyRail commented on code/game/objects/items/weapons/armblade.dm:
```diff
@@ -15,9 +15,104 @@
+	user.delayNextAttack(spin_duration, 1) //Can't attack extra times during the spin
+	for(var/i = 0, i < spin_duration, i++)
+		if(user.incapacitated() || user.locked_to || !isturf(user.loc)) //Double-checking to see if the changeling is allowed to do this
+			user.delayNextMove(-delay_track, 1) //So that the changeling doesn't get magically stuck if it ends early
```
Again, best way to address all these edge cases would be to have add event registers on the most obvious things (like dying, pressing a key, or whatever), have these events change a var on the mob, and during the spin animation check if this var is still valid -- rather than engage the full animation.

ShiftyRail commented on code/game/objects/items/weapons/armblade.dm:
```diff
@@ -15,9 +15,104 @@
+	spin_last_used = world.timeofday
+
+//Returns a value depending on the spin direction and where the user is currently facing.
+/obj/item/weapon/armblade/proc/spin_turn(var/facing, var/direction)
```
Can be as easy as
```dm
turn(facing, -90)
``` 
for left and
```dm
turn(facing, +90)
``` 
for right

ShiftyRail commented:
This seems like a cool idea but surrendering control for 1.5 seconds is a bit gay. I'd recommend adding some way to cancel it early (by moving manually or whatever). This would also cover edge cases like dying midway.

If you wanted to power creep it, you could make the ling immune to stun - or defect electrodes, or whatever - during the spin animation.

boy2mantwicethefam commented:
I don't trust the event system that much because every time I used it it caused some oversight or problem that had to be fixed so I'd rather use regular code checks if it's convenient. I still receive reports that the bomb vest doesn't work as it should despite having 6 different events attached to it.

I don't think the armblade will help much against shotguns and stuns and what-not and I can see the appeal in cancelling it early, but I'd rather it involves using the armblade in hand again rather than move because most people move by reflex and that would mean a lot of early unintentional cancels. The whole thing is balanced with the idea that it's an auto-hitting attack, meaning that the user doesn't have to aim to hit the targets and multiple targets can be hit. There were already a lot of concerns about how OP it might be already so I don't want to touch the spin attack until people are convinced over whether it's actually OP or not.
